File: samples/my_run/all/cpp/simple_live_qml/backend.cpp
/*!
 * \file    backend.cpp
 * \author  IDS Imaging Development Systems GmbH
 * \date    2021-03-24
 * \since   1.0.0
 *
 * \version 1.1.1
 *
 * Copyright (C) 2019 - 2021, IDS Imaging Development Systems GmbH.
 *
 * The information in this document is subject to change without notice
 * and should not be construed as a commitment by IDS Imaging Development Systems GmbH.
 * IDS Imaging Development Systems GmbH does not assume any responsibility for any errors
 * that may appear in this document.
 *
 * This document, or source code, is provided solely as an example of how to utilize
 * IDS Imaging Development Systems GmbH software libraries in a sample application.
 * IDS Imaging Development Systems GmbH does not assume any responsibility
 * for the use or reliability of any portion of this document.
 *
 * General permission to copy or modify is hereby granted.
 */

#include "backend.h"

#include "imageitem.h"

#define VERSION "1.1.2"


BackEnd::BackEnd(QObject* parent)
    : QObject(parent)
{
    // initialize peak library
    peak::Library::Initialize();

    // Create worker thread that waits for new images from the camera
    m_acquisitionWorker = new AcquisitionWorker();
    m_acquisitionWorker->moveToThread(&m_acquisitionThread);

    // worker must be started, when the acquisition starts, and deleted, when the worker thread finishes
    connect(&m_acquisitionThread, &QThread::started, m_acquisitionWorker, &AcquisitionWorker::Start);
    connect(&m_acquisitionThread, &QThread::finished, m_acquisitionWorker, &QObject::deleteLater);

    // connect the worker new image signal with the corresponding backend signal
    connect(m_acquisitionWorker, &AcquisitionWorker::imageReceived, this, &BackEnd::imageChanged);

    // Connect the worker counter updated signal with with the corresponding backend signal
    connect(m_acquisitionWorker, &AcquisitionWorker::counterUpdated, this, &BackEnd::counterChanged);

    // Connect the signal from the acquisition worker when an exception was thrown and a message should be printed
    // with the messagebox trigger slot in the BackEnd class
    connect(m_acquisitionWorker, &AcquisitionWorker::messageBoxTrigger, this, &BackEnd::messageBoxTrigger);
}

BackEnd::~BackEnd()
{
    if (m_acquisitionWorker)
    {
        m_acquisitionWorker->Stop();
        m_acquisitionThread.quit();
        m_acquisitionThread.wait();
    }

    CloseDevice();

    // close peak library
    peak::Library::Close();
}

void BackEnd::OpenDevice()
{
    try
    {
        // Create instance of the device manager
        auto& deviceManager = peak::DeviceManager::Instance();

        // Update the device manager
        deviceManager.Update();

        // Return if no device was found
        if (deviceManager.Devices().empty())
        {
            qDebug() << "[BackEnd::OpenDevice] ERROR: No device found";
            emit messageBoxTrigger("Error", "No device found.");
            return;
        }

        // open the first openable device in the device manager's device list
        size_t deviceCount = deviceManager.Devices().size();
        for (size_t i = 0; i < deviceCount; ++i)
        {
            if (deviceManager.Devices().at(i)->IsOpenable())
            {
                m_device = deviceManager.Devices().at(i)->OpenDevice(peak::core::DeviceAccessType::Control);

                // stop after the first opened device
                break;
            }
            else if (i == (deviceCount - 1))
            {
                qDebug() << "[BackEnd::OpenDevice] ERROR: Device(s) could not be opened";
                emit messageBoxTrigger("Error", "Device could not be openend");
                return;
            }
        }


        if (m_device)
        {
            // Open standard data stream
            m_dataStream = m_device->DataStreams().at(0)->OpenDataStream();

            // Get nodemap of remote device for all accesses to the genicam nodemap tree
            m_nodemapRemoteDevice = m_device->RemoteDevice()->NodeMaps().at(0);

            // To prepare for untriggered continuous image acquisition, load the default user set if available
            // and wait until execution is finished
            try
            {
                m_nodemapRemoteDevice->FindNode<peak::core::nodes::EnumerationNode>("UserSetSelector")
                    ->SetCurrentEntry("Default");
                m_nodemapRemoteDevice->FindNode<peak::core::nodes::CommandNode>("UserSetLoad")->Execute();
                m_nodemapRemoteDevice->FindNode<peak::core::nodes::CommandNode>("UserSetLoad")->WaitUntilDone();
            }
            catch (const peak::core::NotFoundException&)
            {
                // UserSet is not available
            }

            // Get the payload size for correct buffer allocation
            auto payloadSize = m_nodemapRemoteDevice->FindNode<peak::core::nodes::IntegerNode>("PayloadSize")
                                   ->Value();

            // Get the minimum number of buffers that must be announced
            auto bufferCountMax = m_dataStream->NumBuffersAnnouncedMinRequired();

            // Allocate and announce image buffers and queue them
            for (size_t bufferCount = 0; bufferCount < bufferCountMax; ++bufferCount)
            {
                auto buffer = m_dataStream->AllocAndAnnounceBuffer(static_cast<size_t>(payloadSize), nullptr);
                m_dataStream->QueueBuffer(buffer);
            }

            // configure worker
            m_acquisitionWorker->SetDataStream(m_dataStream);
            emit acquisitionStarted();
        }
    }
    catch (const std::exception& e)
    {
        qDebug() << "[BackEnd::StartAcquisition] EXCEPTION: " << e.what();
        emit messageBoxTrigger("Exception", e.what());
        return;
    }

    // Start thread execution
    m_acquisitionThread.start();
}

void BackEnd::CloseDevice()
{
    m_acquisitionWorker->Stop();

    // if device was opened, try to stop acquisition
    if (m_device)
    {
        try
        {
            auto remoteNodeMap = m_device->RemoteDevice()->NodeMaps().at(0);
            remoteNodeMap->FindNode<peak::core::nodes::CommandNode>("AcquisitionStop")->Execute();
        }
        catch (const std::exception& e)
        {
            qDebug() << "[BackEnd::stopAcquisition] EXCEPTION: " << e.what();
            emit messageBoxTrigger("Exception", e.what());
        }
    }

    // if data stream was opened, try to stop it and revoke its image buffers
    if (m_dataStream)
    {
        try
        {
            m_dataStream->KillWait(); //->KillOneWait();
            m_dataStream->StopAcquisition(peak::core::AcquisitionStopMode::Default);
            m_dataStream->Flush(peak::core::DataStreamFlushMode::DiscardAll);

            for (const auto& buffer : m_dataStream->AnnouncedBuffers())
            {
                m_dataStream->RevokeBuffer(buffer);
            }
        }
        catch (const std::exception& e)
        {
            qDebug() << "[BackEnd::StopAcquisition] EXCEPTION: " << e.what();
            emit messageBoxTrigger("Exception", e.what());
        }
    }

    if (m_nodemapRemoteDevice)
    {
        try
        {
            // Unlock parameters after acquisition stop
            m_nodemapRemoteDevice->FindNode<peak::core::nodes::IntegerNode>("TLParamsLocked")->SetValue(0);
        }
        catch (const std::exception& e)
        {
            qDebug() << "[BackEnd::StopAcquisition] EXCEPTION: " << e.what();
            emit messageBoxTrigger("Exception", e.what());
        }
    }
}

QString BackEnd::Version() const
{
    return VERSION;
}

QString BackEnd::QtVersion() const
{
    return qVersion();
}

File: two_cameras_test/BackEnd.cs
/*!
 * \file    BackEnd.cs
 * \author  IDS Imaging Development Systems GmbH
 * \date    2022-06-01
 * \since   1.1.6
 *
 * \version 1.1.0
 *
 * Copyright (C) 2020 - 2023, IDS Imaging Development Systems GmbH.
 *
 * The information in this document is subject to change without notice
 * and should not be construed as a commitment by IDS Imaging Development Systems GmbH.
 * IDS Imaging Development Systems GmbH does not assume any responsibility for any errors
 * that may appear in this document.
 *
 * This document, or source code, is provided solely as an example of how to utilize
 * IDS Imaging Development Systems GmbH software libraries in a sample application.
 * IDS Imaging Development Systems GmbH does not assume any responsibility
 * for the use or reliability of any portion of this document.
 *
 * General permission to copy or modify is hereby granted.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using System.Drawing;
using System.Diagnostics;
using peak.core;

namespace simple_live_windows_forms
{
    class BackEnd
    {
        // Event which is raised if a new image was received
        public delegate void ImageReceivedEventHandler1(object sender, Bitmap image);
        public event ImageReceivedEventHandler1 ImageReceived1;

        // Event which is raised if the counters has changed
        public delegate void CounterChangedEventHandler1(object sender, uint frameCounter, uint errorCounter);
        public event CounterChangedEventHandler1 CounterChanged1;

        // Event which is raised if an Error or Exception has occurred
        public delegate void MessageBoxTriggerEventHandler1(object sender, String messageTitle, String messageText);
        public event MessageBoxTriggerEventHandler1 MessageBoxTrigger1;


        public delegate void ImageReceivedEventHandler2(object sender, Bitmap image);
        public event ImageReceivedEventHandler2 ImageReceived2;

        // Event which is raised if the counters has changed
        public delegate void CounterChangedEventHandler2(object sender, uint frameCounter, uint errorCounter);
        public event CounterChangedEventHandler2 CounterChanged2;

        // Event which is raised if an Error or Exception has occurred
        public delegate void MessageBoxTriggerEventHandler2(object sender, String messageTitle, String messageText);
        public event MessageBoxTriggerEventHandler2 MessageBoxTrigger2;


        private Thread acquisitionThread1;
        private Thread acquisitionThread2;
        private AcquisitionWorker acquisitionWorker1;
        private AcquisitionWorker acquisitionWorker2;

        private Device device1;
        private Device device2;
        private DataStream dataStream1;
        private DataStream dataStream2;
        private NodeMap nodeMapRemoteDevice1;
        private NodeMap nodeMapRemoteDevice2;

        private bool isActive;
        private int opendDevicesCount;


        public BackEnd()
        {
            Debug.WriteLine("--- [BackEnd] Init");

            isActive = true;

            try
            {
                // Create acquisition worker thread that waits for new images from the camera
                acquisitionWorker1 = new AcquisitionWorker(0);
                acquisitionThread1 = new Thread(new ThreadStart(acquisitionWorker1.Start));

                acquisitionWorker1.ImageReceived += acquisitionWorker_ImageReceived1;
                acquisitionWorker1.CounterChanged += acquisitionWorker_CounterChanged1;
                acquisitionWorker1.MessageBoxTrigger += acquisitionWorker_MessageBoxTrigger1;


                acquisitionWorker2 = new AcquisitionWorker(1);
                acquisitionThread2 = new Thread(new ThreadStart(acquisitionWorker2.Start));

                acquisitionWorker2.ImageReceived += acquisitionWorker_ImageReceived2;
                acquisitionWorker2.CounterChanged += acquisitionWorker_CounterChanged2;
                acquisitionWorker2.MessageBoxTrigger += acquisitionWorker_MessageBoxTrigger2;



                // Initialize peak library
                peak.Library.Initialize();
            }
            catch (Exception e)
            {
                Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
            }
        }

        public bool start()
        {
            Debug.WriteLine("--- [BackEnd] Start");
            if (!OpenDevice())
            {
                return false;
            }

            // Start thread execution
            acquisitionThread1.Start();
            acquisitionThread2.Start();

            return true;
        }

        public void Stop()
        {
            Debug.WriteLine("--- [BackEnd] Stop");
            isActive = false;
            acquisitionWorker1.Stop();

            if (acquisitionThread1.IsAlive)
            {
                acquisitionThread1.Join();
            }

            acquisitionWorker2.Stop();

            if (acquisitionThread2.IsAlive)
            {
                acquisitionThread2.Join();
            }


            CloseDevice();

            // Close peak library
            peak.Library.Close();
        }

        public bool OpenDevice()
        {
            Debug.WriteLine("--- [BackEnd] Open device");
            try
            {
                // Create instance of the device manager
                var deviceManager = peak.DeviceManager.Instance();

                // Update the device manager
                deviceManager.Update();

                // Return if no device was found
                if (!deviceManager.Devices().Any())
                {
                    Debug.WriteLine("--- [BackEnd] Error: No device found");
                    MessageBoxTrigger1(this, "Error", "No device found");
                    return false;
                }

                opendDevicesCount = 0;

                // Open the first openable device in the device manager's device list
                var deviceCount = deviceManager.Devices().Count();

                for (var i = 0; i < deviceCount; ++i)
                {
                    if (deviceManager.Devices()[i].IsOpenable())
                    {
                        if (opendDevicesCount == 0)
                            device1 = deviceManager.Devices()[i].OpenDevice(peak.core.DeviceAccessType.Control);
                        else
                            device2 = deviceManager.Devices()[i].OpenDevice(peak.core.DeviceAccessType.Control);

                        opendDevicesCount++;


                        // Stop after the first opened device
                        // break;
                    }
                    else if (i == (deviceCount - 1))
                    {
                        Debug.WriteLine("--- [BackEnd] Error: Device could not be openend");
                        MessageBoxTrigger1(this, "Error", "Device could not be openend");
                        return false;
                    }
                }



                if ((device1 != null)  )
                {
                    // Check if any datastreams are available
                    var dataStreams1 = device1.DataStreams();
                    var dataStreams2 = device2.DataStreams();

                    if (!dataStreams1.Any() & !dataStreams2.Any())
                    {
                        Debug.WriteLine("--- [BackEnd] Error: Device has no DataStream");
                        MessageBoxTrigger1(this, "Error", "Device has no DataStream");
                        return false;
                    }

                    try
                    {
                        // Open standard data stream
                        dataStream1 = dataStreams1[0].OpenDataStream();
                        dataStream2 = dataStreams2[0].OpenDataStream();
                    }
                    catch (Exception e)
                    {
                        Debug.WriteLine("--- [BackEnd] Error: Failed to open DataStream");
                        MessageBoxTrigger1(this, "Error", "Failed to open DataStream\n" + e.Message);
                        return false;
                    }

                    // Get nodemap of remote device for all accesses to the genicam nodemap tree
                    nodeMapRemoteDevice1 = device1.RemoteDevice().NodeMaps()[0];
                    nodeMapRemoteDevice2 = device2.RemoteDevice().NodeMaps()[0];

                    // To prepare for untriggered continuous image acquisition, load the default user set if available
                    // and wait until execution is finished
                    try
                    {
                        nodeMapRemoteDevice1.FindNode<peak.core.nodes.EnumerationNode>("UserSetSelector").SetCurrentEntry("Default");
                        nodeMapRemoteDevice1.FindNode<peak.core.nodes.CommandNode>("UserSetLoad").Execute();
                        nodeMapRemoteDevice1.FindNode<peak.core.nodes.CommandNode>("UserSetLoad").WaitUntilDone();

                        nodeMapRemoteDevice2.FindNode<peak.core.nodes.EnumerationNode>("UserSetSelector").SetCurrentEntry("Default");
                        nodeMapRemoteDevice2.FindNode<peak.core.nodes.CommandNode>("UserSetLoad").Execute();
                        nodeMapRemoteDevice2.FindNode<peak.core.nodes.CommandNode>("UserSetLoad").WaitUntilDone();
                    }
                    catch
                    {
                        // UserSet is not available
                    }

                    // Get the payload size for correct buffer allocation
                    UInt32 payloadSize1 = Convert.ToUInt32(nodeMapRemoteDevice1.FindNode<peak.core.nodes.IntegerNode>("PayloadSize").Value());

                    // Get the minimum number of buffers that must be announced
                    var bufferCountMax1 = dataStream1.NumBuffersAnnouncedMinRequired();

                    // Allocate and announce image buffers and queue them
                    for (var bufferCount = 0; bufferCount < bufferCountMax1; ++bufferCount)
                    {
                        var buffer = dataStream1.AllocAndAnnounceBuffer(payloadSize1, IntPtr.Zero);
                        dataStream1.QueueBuffer(buffer);
                    }

                    // Configure worker
                    acquisitionWorker1.SetDataStream(dataStream1);
                    acquisitionWorker1.SetNodemapRemoteDevice(nodeMapRemoteDevice1);



                    // Get the payload size for correct buffer allocation
                    UInt32 payloadSize2 = Convert.ToUInt32(nodeMapRemoteDevice2.FindNode<peak.core.nodes.IntegerNode>("PayloadSize").Value());

                    // Get the minimum number of buffers that must be announced
                    var bufferCountMax2 = dataStream2.NumBuffersAnnouncedMinRequired();

                    // Allocate and announce image buffers and queue them
                    for (var bufferCount = 0; bufferCount < bufferCountMax2; ++bufferCount)
                    {
                        var buffer = dataStream2.AllocAndAnnounceBuffer(payloadSize2, IntPtr.Zero);
                        dataStream2.QueueBuffer(buffer);
                    }

                    // Configure worker
                    acquisitionWorker2.SetDataStream(dataStream2);
                    acquisitionWorker2.SetNodemapRemoteDevice(nodeMapRemoteDevice2);

                }
            }
            catch (Exception e)
            {
                Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                MessageBoxTrigger1(this, "Exception", e.Message);
                return false;
            }

            return true;
        }

        public void CloseDevice()
        {
            Debug.WriteLine("--- [BackEnd] Close device");
            // If device was opened, try to stop acquisition
            if (device1 != null)
            {
                try
                {
                    var remoteNodeMap1 = device1.RemoteDevice().NodeMaps()[0];
                    remoteNodeMap1.FindNode<peak.core.nodes.CommandNode>("AcquisitionStop").Execute();
                    remoteNodeMap1.FindNode<peak.core.nodes.CommandNode>("AcquisitionStop").WaitUntilDone();
                }
                catch (Exception e)
                {
                    Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                    MessageBoxTrigger1(this, "Exception", e.Message);
                }
            }
            if (device2 != null)
            {
                try
                {
                    var remoteNodeMap2 = device2.RemoteDevice().NodeMaps()[0];
                    remoteNodeMap2.FindNode<peak.core.nodes.CommandNode>("AcquisitionStop").Execute();
                    remoteNodeMap2.FindNode<peak.core.nodes.CommandNode>("AcquisitionStop").WaitUntilDone();
                }
                catch (Exception e)
                {
                    Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                    MessageBoxTrigger1(this, "Exception", e.Message);
                }
            }



            // If data stream was opened, try to stop it and revoke its image buffers
            if (dataStream1 != null)
            {
                try
                {
                    dataStream1.KillWait();
                    dataStream1.StopAcquisition(peak.core.AcquisitionStopMode.Default);
                    dataStream1.Flush(peak.core.DataStreamFlushMode.DiscardAll);

                    foreach (var buffer in dataStream1.AnnouncedBuffers())
                    {
                        dataStream1.RevokeBuffer(buffer);
                    }
                }
                catch (Exception e)
                {
                    Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                    MessageBoxTrigger1(this, "Exception", e.Message);
                }
            }

            if (dataStream2 != null)
            {
                try
                {
                    dataStream2.KillWait();
                    dataStream2.StopAcquisition(peak.core.AcquisitionStopMode.Default);
                    dataStream2.Flush(peak.core.DataStreamFlushMode.DiscardAll);

                    foreach (var buffer in dataStream2.AnnouncedBuffers())
                    {
                        dataStream2.RevokeBuffer(buffer);
                    }
                }
                catch (Exception e)
                {
                    Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                    MessageBoxTrigger1(this, "Exception", e.Message);
                }
            }


            try
            {
                // Unlock parameters after acquisition stop
                nodeMapRemoteDevice1.FindNode<peak.core.nodes.IntegerNode>("TLParamsLocked").SetValue(0);
            }
            catch (Exception e)
            {
                Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                MessageBoxTrigger1(this, "Exception", e.Message);
            }


            try
            {
                // Unlock parameters after acquisition stop
                nodeMapRemoteDevice2.FindNode<peak.core.nodes.IntegerNode>("TLParamsLocked").SetValue(0);
            }
            catch (Exception e)
            {
                Debug.WriteLine("--- [BackEnd] Exception: " + e.Message);
                MessageBoxTrigger1(this, "Exception", e.Message);
            }

        }

        private void acquisitionWorker_ImageReceived1(object sender, System.Drawing.Bitmap image)
        {
            ImageReceived1(sender, image);
        }

        private void acquisitionWorker_CounterChanged1(object sender, uint frameCounter, uint errorCounter)
        {
            CounterChanged1(sender, frameCounter, errorCounter);
        }

        private void acquisitionWorker_MessageBoxTrigger1(object sender, String messageTitle, String messageText)
        {
            MessageBoxTrigger1(sender, messageTitle, messageText);
        }


        private void acquisitionWorker_ImageReceived2(object sender, System.Drawing.Bitmap image)
        {
            ImageReceived2(sender, image);
        }

        private void acquisitionWorker_CounterChanged2(object sender, uint frameCounter, uint errorCounter)
        {
            CounterChanged2(sender, frameCounter, errorCounter);
        }

        private void acquisitionWorker_MessageBoxTrigger2(object sender, String messageTitle, String messageText)
        {
            MessageBoxTrigger2(sender, messageTitle, messageText);
        }


        public bool IsActive()
        {
            return isActive;
        }
    }
}
